### ISSUE TITLE
[Feature request] Pass custom `fetch` function

### ISSUE THREAD

**Issue by KaKi87** [haverstack/axios-fetch-adapter#6]
Hello,

It would be nice to be able to pass a specific `fetch` function to use the adapter with, so that it could support peculiar environments such as [SvelteKit's SSR `load` function](https://kit.svelte.dev/docs/load#making-fetch-requests).

Thanks

**Comment by KaKi87** [comment]
@cuibonobo 🙏

**Comment by cuibonobo** [comment]
What do you propose the API for this should be?

**Comment by KaKi87** [comment]
How about `axios.create({ adapter: fetchAdapter({ fetch }) });` ? Thanks

**Comment by cuibonobo** [comment]
I was afraid you would say this! I have been testing out this API and it might be dangerous if Axios ever changes their config params to include `"fetch"`. The `fetchAdapter()` function needs to return a `Promise` when it's used inside of Axios, but with the proposed API, it would need to return either a `Promise` or a function, depending on whether you pass a `fetch` function or not. If Axios changes their config in the future to include a `"fetch"` param, it will clash with this API so it will not be clear whether to return the `Promise` or the function.

That was a mouthful, so I'll illustrate with some code:
```typescript
// Signature for the original `fetchAdapter` function. Axios uses this internally to make requests. It accepts
// an `AxiosRequestConfig` object and returns a `Promise`
function fetchAdapter(config: AxiosRequestConfig): Promise<AxiosResponse>;

// Proposed API, which *generates* a function with the above signature. Luckily `AxiosRequestConfig`
// from above does not contain a "fetch" key, but if that ever changed in the future, we wouldn't be
// able to distinguish the params between the top signature and the bottom signature.
function fetchAdapter(config: {fetch: any}): (config: AxiosRequestConfig) => Promise<AxiosResponse>;

```

**Comment by cuibonobo** [comment]
Maybe an alternative could be a separate function for setting a custom `fetch` if that is ever needed? For example:
```typescript
import axios from "axios";
import myFetch from "myFetch";
import fetchAdapter, { fetchAdapterFactory } from "@haverstack/axios-fetch-adapter";

// Current behavior
const defaultFetchClient = axios.create({
  adapter: fetchAdapter
});

// Proposal for custom `fetch` functions
const customFetchClient = axios.create({
  adapter: fetchAdapterFactory(myFetch)
});
```

**Comment by KaKi87** [comment]
Actually, I was implicitly suggesting that `fetchAdapter` would, from now on, permanently use the second signature, which, of course, would be a backwards-incompatible change.

The lib still being in prerelease, it shouldn't be a problem, right ?

**Comment by cuibonobo** [comment]
I'm not willing to do that.

**Comment by KaKi87** [comment]
Why not ?

**Comment by cuibonobo** [comment]
I don't want to introduce a backwards-incompatible change.

**Comment by KaKi87** [comment]
Okay, then :

> with the proposed API, it would need to return either a `Promise` or a function, depending on whether you pass a `fetch` function or not. If Axios changes their config in the future to include a `"fetch"` param, it will clash with this API

Here's an alternative : instead of checking if something for the adapter is passed, check whether something for Axios is passed (e.g. `url`).

What do you think about that ?

**Comment by cuibonobo** [comment]
I don't see why these functions have to be called the same thing. What's wrong with using a factory function when you want to set your own `fetch`?

**Comment by KaKi87** [comment]
That's an option as well. I just find the first one more sexy, but if you choose the second one then I'll go with it.

But in that case, please drop the OOP vocabulary that JS doesn't need and just call it `createFetchAdapter`.

Also, here's an intermediate alternative : still using `fetchAdapter()`, but with `{ adapterConfig: { fetch } }` as parameter, preventing any future ambiguity.

**Comment by KaKi87** [comment]
Oh and regarding the second option, I just noticed that you're proposing to pass `fetch` without putting it in an object : I don't recommend this, because it makes it less convenient to add future configuration options if needed.

Thanks

**Comment by cuibonobo** [comment]
I'm not really interested in sexyness, but I appreciate the alternative name suggestion.

**Comment by cuibonobo** [comment]
Fixed in 4bd2d4984b88679eb36ca22f20a256d7df0a0916. Deployed to NPM as [v0.8.0](https://www.npmjs.com/package/@haverstack/axios-fetch-adapter/v/0.8.0).

**Comment by KaKi87** [comment]
Thanks !

However, I've got a new problem : it returns `ERR_INVALID_URL` when using an `url` that starts with `/`, which, using normal `fetch`, would indeed be incorrect, but using a custom one in an SSR context, is proper.

Thanks

**Comment by cuibonobo** [comment]
Could setting the `baseURL` in the Axios config to `http://localhost` help with this?

**Comment by KaKi87** [comment]
No, because then it would make an actual request to `localhost`, which will obviously fail.

**Comment by cuibonobo** [comment]
I'm honestly not sure how to accommodate this. The `ERR_INVALID_URL` is coming from Axios itself, not the code in this repo. But even if that weren't the case, this adapter works by creating a new `Request` object, which expects a fully-qualified URL. Does Svelte have a custom `Request` object as well? If you can get Axios to stop yelling at you, maybe the fix could be something like:

```javascript
const myFetchAdapter = createFetchAdapter({ fetch: myFetch, request: myRequest });
```

**Comment by KaKi87** [comment]
Here's the stacktrace :

```
TypeError: Failed to parse URL from /some-endpoint
    at new Request (node_modules/undici/lib/fetch/request.js:87:15)
    at createRequest (node_modules/@haverstack/axios-fetch-adapter/lib/index.js:440:10)
    ... 5 lines matching cause stack trace ...
    at axiosAdapter (node_modules/@haverstack/axios-fetch-adapter/lib/index.js:336:12)
    at Axios.dispatchRequest (file://node_modules/axios/lib/core/dispatchRequest.js:51:10)
    at Axios.request (file://node_modules/axios/lib/core/Axios.js:148:33) {
  [cause]: TypeError [ERR_INVALID_URL]: Invalid URL
      at new NodeError (node:internal/errors:405:5)
      at new URL (node:internal/url:611:13)
      at new Request (node_modules/undici/lib/fetch/request.js:85:21)
      at createRequest (node_modules/@haverstack/axios-fetch-adapter/lib/index.js:440:10)
      at node_modules/@haverstack/axios-fetch-adapter/lib/index.js:337:23
      at Generator.next (<anonymous>)
      at node_modules/@haverstack/axios-fetch-adapter/lib/index.js:66:61
      at new Promise (<anonymous>)
      at __async (node_modules/@haverstack/axios-fetch-adapter/lib/index.js:50:10)
      at axiosAdapter (node_modules/@haverstack/axios-fetch-adapter/lib/index.js:336:12) {
    input: '/some-endpoint',
    code: 'ERR_INVALID_URL'
  }
}
```

Which seems to indicate that the issue comes from `Request` indeed.

Could the usage of `Request` be disabled when passing a custom `fetch` instance ?

Thanks

**Comment by cuibonobo** [comment]
The `Request` object is useful for building up authentication headers and whatnot. Better to disable it explicitly if it's not desired, because of the possibly strange side-effects. Something like:

```javascript
const myFetchAdapter = createFetchAdapter({ fetch: myFetch, disableRequest: true });
```

What do you think?

**Comment by KaKi87** [comment]
> The `Request` object is useful for building up authentication headers and whatnot.

Why ? The official adapters don't need it, e.g. here's what their XHR one uses :

https://github.com/axios/axios/blob/3f53eb6960f05a1f88409c4b731a40de595cb825/lib/adapters/xhr.js#L74-L79

Also, I made an adapter myself for another environment, so I used the same idea :

https://git.kaki87.net/KaKi87/axios-tauri-adapter/src/branch/master/index.js#L43

> Better to disable it explicitly if it's not desired, because of the possibly strange side-effects.

I'm afraid such effects are unacceaptable.

Thanks

**Comment by cuibonobo** [comment]
The idea is to pass the URL directly to the `fetch` function without creating a `Request` object first. An example of a side effect of this is that the `AxiosResponse` object will just have this URL in its `request` property. This means that accessing, for example, `response.request.url` won't work.

**Comment by KaKi87** [comment]
Well, if only `response.request` is affected then it's fine.

But if `config` is affected then its not.

**Comment by KaKi87** [comment]
What's the verdict ?

Thanks

**Comment by KaKi87** [comment]
Hello ?

**Comment by KaKi87** [comment]
Hello,
Any news on this ?
Thanks

**Comment by cuibonobo** [comment]
The `disableRequest` option is now available in version 0.11.0 on NPM.
